platforms/extensibility/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java

```diff
@@ -285,19 +252,15 @@ public void found(String sourceDescription, PluginResolution pluginResolution) {
             found = pluginResolution;
         }
 
-        @Override
-        public void alreadyApplied() {
-            alreadyApplied = true;
-        }
-
         @Override
         public boolean isFound() {
-            return alreadyApplied || found != null;
+            return found != null;
         }
     }
 
     private static class PluginDependencyVisitor implements PluginResolveContext {
         private List<Dependency> additionalDependencies;
+        private List<Pair<ModuleIdentifier, ModuleIdentifier>> replacements;
```

Comment:
🤔 A list of `Pair`s means you have to keep track of what is to the left and right when you read this data in `DefaultPluginRequestApplicator`.  If this were a `Map<ModuleIdentifier, ModuleIdentifier>`, I think the code accessing this collection would be more self-explanatory.